File: sdk/php/src/Internal/Ws/DefaultKucoinWsImpl.php
<?php

namespace KuCoin\UniversalSDK\Internal\Ws;

use KuCoin\UniversalSDK\Api\KucoinWSService;
use KuCoin\UniversalSDK\Generate\Futures\FuturesPrivate\FuturesPrivateWs;
use KuCoin\UniversalSDK\Generate\Futures\FuturesPrivate\FuturesPrivateWsImpl;
use KuCoin\UniversalSDK\Generate\Futures\FuturesPublic\FuturesPublicWs;
use KuCoin\UniversalSDK\Generate\Futures\FuturesPublic\FuturesPublicWsImpl;
use KuCoin\UniversalSDK\Generate\Margin\MarginPrivate\MarginPrivateWs;
use KuCoin\UniversalSDK\Generate\Margin\MarginPrivate\MarginPrivateWsImpl;
use KuCoin\UniversalSDK\Generate\Margin\MarginPublic\MarginPublicWs;
use KuCoin\UniversalSDK\Generate\Margin\MarginPublic\MarginPublicWsImpl;
use KuCoin\UniversalSDK\Generate\Spot\SpotPrivate\SpotPrivateWs;
use KuCoin\UniversalSDK\Generate\Spot\SpotPrivate\SpotPrivateWsImpl;
use KuCoin\UniversalSDK\Generate\Spot\SpotPublic\SpotPublicWs;
use KuCoin\UniversalSDK\Generate\Spot\SpotPublic\SpotPublicWsImpl;
use KuCoin\UniversalSDK\Generate\Version;
use KuCoin\UniversalSDK\Internal\Infra\DefaultWsService;
use KuCoin\UniversalSDK\Model\ClientOption;
use KuCoin\UniversalSDK\Model\Constants;
use React\EventLoop\Loop;
use React\EventLoop\LoopInterface;

class DefaultKucoinWsImpl implements KucoinWSService
{
    /**
     * @var ClientOption $clientOption
     */
    private $clientOption;


    /**
     * @var LoopInterface $loop
     */
    private $loop;

    public function __construct(ClientOption $clientOption, ?LoopInterface $loop = null)
    {
        $this->clientOption = $clientOption;
        $this->loop = $loop ?: Loop::get();
    }

    public function startEventLoop()
    {
        $this->loop->run();
    }

    public function stopEventLoop()
    {
        $this->loop->stop();
    }

    public function getLoop(): LoopInterface
    {
        return $this->loop;
    }


    public function newSpotPublicWS(): SpotPublicWs
    {
        $wsService = new DefaultWsService(
            $this->clientOption,
            $this->loop,
            Constants::DOMAIN_TYPE_SPOT,
            false,
            Version::SDK_VERSION);

        return new SpotPublicWsImpl($wsService);
    }

    public function newSpotPrivateWS(): SpotPrivateWs
    {
        $wsService = new DefaultWsService(
            $this->clientOption,
            $this->loop,

            Constants::DOMAIN_TYPE_SPOT,
            true,
            Version::SDK_VERSION);

        return new SpotPrivateWsImpl($wsService);
    }

    public function newMarginPublicWS(): MarginPublicWs
    {
        $wsService = new DefaultWsService(
            $this->clientOption,
            $this->loop,
            Constants::DOMAIN_TYPE_SPOT,
            false,
            Version::SDK_VERSION);

        return new MarginPublicWsImpl($wsService);
    }

    public function newMarginPrivateWS(): MarginPrivateWs
    {
        $wsService = new DefaultWsService(
            $this->clientOption,
            $this->loop,
            Constants::DOMAIN_TYPE_SPOT,
            true,
            Version::SDK_VERSION);

        return new MarginPrivateWsImpl($wsService);
    }

    public function newFuturesPublicWS(): FuturesPublicWs
    {
        $wsService = new DefaultWsService(
            $this->clientOption,
            $this->loop,
            Constants::DOMAIN_TYPE_FUTURES,
            false,
            Version::SDK_VERSION);

        return new FuturesPublicWsImpl($wsService);
    }

    public function newFuturesPrivateWS(): FuturesPrivateWs
    {
        $wsService = new DefaultWsService(
            $this->clientOption,
            $this->loop,
            Constants::DOMAIN_TYPE_FUTURES,
            true,
            Version::SDK_VERSION);

        return new FuturesPrivateWsImpl($wsService);
    }
}

File: sdk/java/src/main/java/com/kucoin/universal/sdk/internal/ws/DefaultKucoinWsImpl.java
package com.kucoin.universal.sdk.internal.ws;

import com.kucoin.universal.sdk.api.KucoinWSService;
import com.kucoin.universal.sdk.generate.Version;
import com.kucoin.universal.sdk.generate.futures.futuresprivate.FuturesPrivateWs;
import com.kucoin.universal.sdk.generate.futures.futuresprivate.FuturesPrivateWsImpl;
import com.kucoin.universal.sdk.generate.futures.futurespublic.FuturesPublicWs;
import com.kucoin.universal.sdk.generate.futures.futurespublic.FuturesPublicWsImpl;
import com.kucoin.universal.sdk.generate.margin.marginprivate.MarginPrivateWs;
import com.kucoin.universal.sdk.generate.margin.marginprivate.MarginPrivateWsImpl;
import com.kucoin.universal.sdk.generate.margin.marginpublic.MarginPublicWs;
import com.kucoin.universal.sdk.generate.margin.marginpublic.MarginPublicWsImpl;
import com.kucoin.universal.sdk.generate.spot.spotprivate.SpotPrivateWs;
import com.kucoin.universal.sdk.generate.spot.spotprivate.SpotPrivateWsImpl;
import com.kucoin.universal.sdk.generate.spot.spotpublic.SpotPublicWs;
import com.kucoin.universal.sdk.generate.spot.spotpublic.SpotPublicWsImpl;
import com.kucoin.universal.sdk.internal.infra.DefaultWsService;
import com.kucoin.universal.sdk.model.ClientOption;
import com.kucoin.universal.sdk.model.Constants;

/** DefaultKucoinWsImpl provides WebSocket interfaces for Spot, Margin, and Futures trading. */
public final class DefaultKucoinWsImpl implements KucoinWSService {

  /** Client configuration options. */
  private final ClientOption clientOption;

  public DefaultKucoinWsImpl(ClientOption clientOption) {
    this.clientOption = clientOption;
  }

  /**
   * Returns the interface to interact with the Spot Trading WebSocket (public channel) API of
   * KuCoin.
   *
   * @return SpotPublicWs
   */
  @Override
  public SpotPublicWs newSpotPublicWS() {
    DefaultWsService wsService =
        new DefaultWsService(clientOption, Constants.DOMAIN_TYPE_SPOT, false, Version.SDK_VERSION);
    return new SpotPublicWsImpl(wsService);
  }

  /**
   * Returns the interface to interact with the Spot Trading WebSocket (private channel) API of
   * KuCoin.
   *
   * @return SpotPrivateWs
   */
  @Override
  public SpotPrivateWs newSpotPrivateWS() {
    DefaultWsService wsService =
        new DefaultWsService(clientOption, Constants.DOMAIN_TYPE_SPOT, true, Version.SDK_VERSION);
    return new SpotPrivateWsImpl(wsService);
  }

  /**
   * Returns the interface to interact with the Margin Trading WebSocket (public channel) API of
   * KuCoin.
   *
   * @return MarginPublicWs
   */
  @Override
  public MarginPublicWs newMarginPublicWS() {
    DefaultWsService wsService =
        new DefaultWsService(clientOption, Constants.DOMAIN_TYPE_SPOT, false, Version.SDK_VERSION);
    return new MarginPublicWsImpl(wsService);
  }

  /**
   * Returns the interface to interact with the Margin Trading WebSocket (private channel) API of
   * KuCoin.
   *
   * @return MarginPrivateWs
   */
  @Override
  public MarginPrivateWs newMarginPrivateWS() {
    DefaultWsService wsService =
        new DefaultWsService(clientOption, Constants.DOMAIN_TYPE_SPOT, true, Version.SDK_VERSION);
    return new MarginPrivateWsImpl(wsService);
  }

  /**
   * Returns the interface to interact with the Futures Trading WebSocket (public channel) API of
   * KuCoin.
   *
   * @return FuturesPublicWs
   */
  @Override
  public FuturesPublicWs newFuturesPublicWS() {
    DefaultWsService wsService =
        new DefaultWsService(
            clientOption, Constants.DOMAIN_TYPE_FUTURES, false, Version.SDK_VERSION);
    return new FuturesPublicWsImpl(wsService);
  }

  /**
   * Returns the interface to interact with the Futures Trading WebSocket (private channel) API of
   * KuCoin.
   *
   * @return FuturesPrivateWs
   */
  @Override
  public FuturesPrivateWs newFuturesPrivateWS() {
    DefaultWsService wsService =
        new DefaultWsService(
            clientOption, Constants.DOMAIN_TYPE_FUTURES, true, Version.SDK_VERSION);
    return new FuturesPrivateWsImpl(wsService);
  }
}
